corgi/tasks/yum.py

```diff
@@ -0,0 +1,63 @@
+import logging
+from urllib.parse import urlparse
+
+from celery_singleton import Singleton
+
+from config.celery import app
+from corgi.collectors.yum import Yum
+from corgi.core.models import Channel, ProductComponentRelation, ProductStream
+from corgi.tasks.common import RETRY_KWARGS, RETRYABLE_ERRORS, _create_relations
+
+logger = logging.getLogger(__name__)
+
+
+@app.task(base=Singleton, autoretry_for=RETRYABLE_ERRORS, retry_kwargs=RETRY_KWARGS)
+def load_yum_repositories() -> None:
+    """Use dnf repoquery commands to inspect and load all content in all Yum repos"""
+    logger.info("Loading all Yum repository data for all ProductStreams")
+    for stream, repos in ProductStream.objects.exclude(yum_repositories=[]).values_list(
+        "name", "yum_repositories"
+    ):
+        for repo in repos:  # type: ignore
+            # mypy thinks yum_repositories can be None, even though it defaults to []
+            slow_load_yum_repositories_for_stream.delay(stream, repo)
+
+
+@app.task(base=Singleton, autoretry_for=RETRYABLE_ERRORS, retry_kwargs=RETRY_KWARGS)
+def slow_load_yum_repositories_for_stream(stream: str, repo: str) -> None:
+    """Use dnf repoquery commands to inspect and load all content in a particular Yum repo"""
+    logger.info(f"Loading Yum repository {repo} for ProductStream {stream}")
+
+    # Some "Yum repository" URLs in prod-defs are actually Pulp repo URLs
+    # Check to see if we already made a Channel for them, by looking up the repo's relative URL
+    # If not, create the repo using the full path in prod-defs as the name
+    logger.info(f"Checking if Channel with matching URL already exists for {repo}")
+    relative_url = urlparse(repo).path.lstrip("/")
+    channel, created = Channel.objects.get_or_create(
+        relative_url=relative_url, defaults={"name": repo}
+    )
+
+    if created:
+        logger.info(f"Created new Channel for {repo}")
+    else:
+        logger.info(f"Found existing Channel for {repo} with name {channel.name}")
+
+    logger.info(f"Querying Yum repository {repo} for SRPMs")
+    yum = Yum()
+    srpm_build_ids = yum.get_srpms_from_yum_repos((repo,))
+    # Save the Brew build IDs for all SRPMs in the repo into the ProductComponentRelation table
+    # The weekly "fetch_unprocessed_cdn_relations" Celery task will look up these IDs
+    # then create Components and SoftwareBuilds for each (including both RPMs and modules)
+    srpm_relations = _create_relations(
+        srpm_build_ids, channel.name, stream, ProductComponentRelation.Type.CDN_REPO
+    )
+    if srpm_relations > 0:
+        logger.info(f"Created {srpm_relations} new relations for SRPMs in {channel.name}")
+
+    logger.info(f"Querying Yum repository {repo} for modules")
+    module_build_ids = yum.get_modules_from_yum_repos((repo,))
+    module_relations = _create_relations(
+        module_build_ids, channel.name, stream, ProductComponentRelation.Type.CDN_REPO
```

Comment:
This will cause issues with SoftwareBuild.save_product_taxonomy(). In that function we expect CDN_REPO types to be associated with variants, not streams. Perhaps it's better to use a new type YUM_REPO and link it with a stream instead, so that we know to expect a stream name always for YUM_REPOS.

Reply:
An alternative, more invasive fix is to use ofuri in product_ref column of relations table. When reading relations back from the table we could then parse the ofuri to determine the product type it refers to.

Reply:
Ack, in that case I'll switch the type of the created relations to YUM_REPO, then copy the Pulp task code that finds unprocessed CDN relations to the Yum task module and make it look for unprocessed YUM relations instead.